ansible_base/lib/utils/models.py

```diff
@@ -97,3 +99,101 @@ def current_user_or_system_user():
     if user is None or user.is_anonymous:
         user = get_system_user()
     return user
+
+
+def diff(old, new, require_type_match=True, json_safe=True, include_m2m=False, exclude_fields=[], limit_fields=[]):
+    """
+    Diff two instances of models (which do not have to be the same type of model
+    if given require_type_match=False).
+
+    This function is used in particular by the activitystream application where
+    the changes returned by this function are stored as models change.
+
+    :param old: The old instance for comparison
+    :param new: The new instance for comparison
+    :param require_type_match: If True, old and new must be of the same type of
+        model. (default: True)
+    :param json_safe: If True, the diff will be made JSON-safe by converting
+        all non-JSON-safe values to strings using Django's smart_str function.
+        (default: True)
+    :param exclude_fields: A list of field names to exclude from the diff.
+        (default: [])
+    :param limit_fields: A list of field names to limit the diff to. This can be
+        useful, for example, when update_fields is passed to a model's save
+        method and you only want to diff the fields that were updated.
+        (default: [])
+    :param include_m2m: If True, include many-to-many fields in the diff.
+        Otherwise, they are ignored. (default: False)
+    :return: A dictionary with the following
+        - added_fields: A dictionary of fields that were added between old and
+          new. Importantly, if old and new are the same type of model, this
+          should always be empty. An "added field" does not mean that the field
+          became non-empty, it means that the field was completely absent from
+          the old type of model and is now present in the new type of model. If
+          this entry is non-empty, it has the form: {"field_name": value} where
+          value is the new value of the field.
+        - removed_fields: A dictionary of fields that were removed between old
+          and new. Similar to added_fields, if old and new are the same type of
+          model, this should always be empty. It has the same structure as
+          added_fields except the value is the old value of the field.
+        - changed_fields: A dictionary of fields that were changed between old
+          and new. The key of each entry is the field name, and the value is a
+          tuple of the old value and the new value.
+    """
+
+    def skip_field(source, field):
+        if field in exclude_fields:
+            return True
+        if limit_fields and field not in limit_fields:
+            return True
+        if not include_m2m and source._meta.get_field(field).many_to_many:
+            return True
+        return False
```

Comment:
prefer that `def` is moved out of this method.

Reply:
I don't think this inner function is very useful as a general function, and it's only because I was repeating this logic in this function a lot, so I wanted to make it less annoying to read and write.